src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,1070 @@
+    // A bit that is 0 in both lo and zeros must be 0 in either
+    //           1 0 0 1 1 1 1 0
+    U either = lo | bits._zeros;
+    // This is all the bits that are not lower than first_violation and are 0
+    // in both lo and zeros. Note that there may not exist such bits. In those
+    // cases tmp == 0 and there is no value not less than lo and satisfies
+    // bits. We will expand the implication of such cases below.
+    //           0 1 1 0 0 0 0 0
+    U tmp = ~either & find_mask;
+    // According to the conclusion in the overview of this case above, i is the
+    // last bit being 0 in both lo and zeros that is upto the first violation,
+    // which is the last set bit of tmp.
+    // In our example, i == 2 here, we want to obtain the value with only the
+    // bit i set, this is equivalent to extracting the last set bit of tmp, do
+    // it directly without going through i.
+    //           0 0 1 0 0 0 0 0
+    U alignment = tmp & (-tmp);
+    // Set the bit of lo at i and unset all the bits after, this is the smallest
+    // value that satisfies bits._zeros. Similar to the above case, this is
+    // similar to aligning lo upto alignment. Also similar to the above case,
+    // this computation cannot overflow.
+    // We now have:
+    // - new_lo[x] = lo[x], for 0 <= x < i (2.5)
+    // - new_lo[i] = 1                     (2.6)
+    // - new_lo[x] = 0, for x > i          (not yet 2.7)
+    //           1 0 1 0 0 0 0 0
+    U new_lo = (lo & -alignment) + alignment;
+    // Note that there exists no value x not larger than i such that
+    // new_lo[x] == 0 and ones[x] == 1. This is because all bits of lo before i
+    // should satisfy bits, and new_lo[i] == 1. As a result, doing
+    // new_lo |= bits.ones will give us a value such that:
+    // - new_lo[x] = lo[x], for 0 <= x < i (2.5)
+    // - new_lo[i] = 1                     (2.6)
+    // - new_lo[x] = ones[x], for x > i    (2.7)
+    // This is the result we are looking for.
+    //           1 0 1 0 0 0 1 1
+    new_lo |= bits._ones;
+    // In this case, new_lo may not always be a valid answer. This can happen
+    // if there is no bit upto first_violation that is 0 in both lo and zeros,
+    // i.e. tmp == 0. In such cases, alignment == 0 && lo == bits._ones. It is
+    // the only case when this function does not return a valid answer.
+    // Note: it can be seen as an overflow because we can say that the
+    // computation of tmp above results in 0b...1100..00 with W trailing 0s.
+    // As a result, alignment should be 0b100...00 with W trailing 0s. This
+    // overflows the W-bit arithmetic and we obtain the value alignment == 0.
+    // We can say that the algorithm should round up lo to a multiple of 2**W.
+    // This overflows and the highest bit is discarded, leaving us with the
+    // result of rounding up being 0.
+    assert(lo < new_lo || new_lo == bits._ones, "overflow must return bits._ones");
+    return new_lo;
```

Comment:
Hmm, I would not say that it can return a **non valid** answer. Because we declared that it is ok to return overflowed values at the top of the method, in fact we must prove that in those cases where we cannot find an `r` we at least get `new_lo < lo`.